### ISSUE TITLE
regression with tokio-openssl and large frame sizes (0.13 -> 0.14)

### ISSUE THREAD

**Issue by Fabian-Gruenbichler** [hyperium/hyper#2389]
in the process of upgrading from hyper 0.13 (h2 0.2/tokio 0.2/tokio-openssl 0.4) to 0.14 (h2 0.3/tokio 1.0/tokio-openssl 0.6) I noticed the following regression in performance when downloading data via http2+TLS with a large `max_frame_size` set (although I suspect this exact setup is not the only thing affected). the issue seems to boil down to something causing big, uninitialized `ReadBuf`s being passed down the stack when the (max) frame size is big. tokio-openssl (which wraps the non-async openssl crate, which in turns binds to libssl/libcrypto) then fully initializes this big ReadBuf to pass the resulting slice to a regular `Read` implementation - on each poll_read!

the regression is rather hefty, when (serially) making 2000 requests for 4MB of data each a frame size of 512k (instead of the default 16k) already makes the test case ~1.5x slower, 4M bumps that even further to 7x (compared to the old stack with the same frame size). non-TLS (completely async) performance is not affected:

```
Benchmark #11: cargo run --release --bin h2s-client 512
  Time (mean ± σ):      9.647 s ±  0.025 s    [User: 8.222 s, System: 2.744 s]
  Range (min … max):    9.629 s …  9.665 s    2 runs

Benchmark #12: cargo run --release --bin baseline-h2s-client 512
  Time (mean ± σ):      6.422 s ±  0.019 s    [User: 4.991 s, System: 3.520 s]
  Range (min … max):    6.408 s …  6.435 s    2 runs

...

Benchmark #17: cargo run --release --bin h2s-client 4096
  Time (mean ± σ):     43.277 s ±  0.029 s    [User: 42.394 s, System: 2.150 s]
  Range (min … max):   43.256 s … 43.297 s    2 runs

Benchmark #18: cargo run --release --bin baseline-h2s-client 4096
  Time (mean ± σ):      6.389 s ±  0.016 s    [User: 5.052 s, System: 3.401 s]
  Range (min … max):    6.378 s …  6.400 s    2 runs
```

I pushed a reduced proof of concept and some benchmark/flamegraph data here:
https://github.com/Fabian-Gruenbichler/hyper-regression

the flamegraph gives a good overview over what's going on.

I am not sure who's at fault here or how to improve/fix this (except the obvious workaround of not setting such big frame sizes ;)), but my guess is tokio-openssl is not the only thing that needs to initialize passed-in ReadBuf's that is used together with hyper, so this might affect other people as well..

tagging @sfackler as well in case they want to chime in with a (tokio-)openssl perspective

**Comment by sfackler** [comment]
I believe Hyper should be remembering the initialization state of its buffers, though tokio-openssl should be able to pass the uninitialized buffers directly to OpenSSL if I add some extra methods to SslStream.

**Comment by seanmonstar** [comment]
For h2, hyper is utilizing `tokio_util::codec`, so the fix would need to be in there. Looks like it does so [generically](https://github.com/tokio-rs/tokio/blob/e42317b072c5240fe190c0a8708cd24bb4a4eb08/tokio-util/src/lib.rs#L106), so it doesn't know how much of the `BufMut` internally has been initialized...

**Comment by sfackler** [comment]
https://github.com/sfackler/tokio-openssl/commit/65e5c482bdb9cdca151430f7970f8c988ddf3982 should fix the overhead, but this is a tokio-util bug at its root.

**Comment by Blub** [comment]
`ReadBuf` probably introduces quite a few such issues. Eg. [Buffered::poll_read_from_io] creates a `ReadBuf::uninit` over the whole `BytesMut` chunk. If the underlying `io` object is one which bridges from a `ReadBuf`-using code to a non-`ReadBuf` code (eg. `tokio-openssl`), it'll initialize the entire buffer to then pass it to the non-`ReadBuf` code.
If this yields, say, 3 bytes with the `read_buf` being eg. 1024 bytes, the next call will be creating a `ReadBuf::uninit` over the remaining 1021 bytes and the underling `io` code will re-initialize all of those already-initialized 1021 bytes.

Not sure where exactly the above example slows down, but IMO `Buffered` does need to remember the initialized range of the created `ReadBuf` since currently it just discards that information.

Long story short, any code which creates, passes along, and then *drops* an uninit `ReadBuf` should probably consider remember the `ReadBuf`'s state for later instead.

[Buffered::poll_read_from_io]: https://github.com/hyperium/hyper/blob/c9c46ed60bd43484d67a98a63496ad0277c791a2/src/proto/h1/io.rs#L192

**Comment by Fabian-Gruenbichler** [comment]
> [sfackler/tokio-openssl@65e5c48](https://github.com/sfackler/tokio-openssl/commit/65e5c482bdb9cdca151430f7970f8c988ddf3982) should fix the overhead, but this is a tokio-util bug at its root.

thanks for the fast turn-around! :) 0.6.1 fixes the issue (both for the reproducer and the original triggering code).

it would still be nice to get a proper fix in tokio-util (and maybe a note going in the direction of @Blub 's comment for `ReadBuf::uninit`'s documentation?) - do either of you (@sfackler / @seanmonstar) want to do the forwarding or should I?

**Comment by seanmonstar** [comment]
If you're motivated, you're welcome to champion it!

Since it's partially fixed and not something hyper itself can change, I'm going to close this.

**Comment by Fabian-Gruenbichler** [comment]
> If you're motivated, you're welcome to champion it!
> 
> Since it's partially fixed and not something hyper itself can change, I'm going to close this.

filed: https://github.com/tokio-rs/tokio/issues/3417

isn't the code @Blub linked also affected though? the `read_buf` `BytesMut` might not change between calls, but the derived `ReadBuf` is always assumed to be fully uninitialized (and thus potentially re-initialized by the `AsyncRead` implementation). ideally we'd keep track of the initialization state and `assume_init` accordingly. not sure how hot that code path is in practice..

**Comment by Darksonn** [comment]
I'll just comment here too, but this cannot be fixed in `tokio-util` as it would be unsound to do so. This is because user code could de-initialize the buffer between calls to `poll_read`.
